core/src/main/java/io/apicurio/hub/api/codegen/OpenApi2JaxRs.java

```diff
@@ -718,6 +748,56 @@ protected static String toStringArrayLiteral(Set<String> values) {
         return builder.toString();
     }
 
+    void addValidationConstraints(AnnotationTargetSource<?, ?> target, CodegenJavaSchema schemaInfo, boolean forbidNotNull, String topLevelPackage) {
+        final String constraintPkg = String.format("%s.validation.constraints", topLevelPackage);
+        final String sizeConstraint = String.format("%s.Size", constraintPkg);
+
+        if (!forbidNotNull && !Boolean.TRUE.equals(schemaInfo.getNullable())) {
+            // nullable is false by default
+            target.addAnnotation(String.format("%s.validation.constraints.NotNull", topLevelPackage));
+        }
+
+        if (schemaInfo.getCollection() != null) {
+            if ("map".equals(schemaInfo.getCollection())) {
+                addConstraint(target, schemaInfo.getMaxProperties(), sizeConstraint, "max");
+                addConstraint(target, schemaInfo.getMinProperties(), sizeConstraint, "min");
+            } else {
+                addConstraint(target, schemaInfo.getMaxItems(), sizeConstraint, "max");
+                addConstraint(target, schemaInfo.getMinItems(), sizeConstraint, "min");
+            }
+            return;
+        }
+
+        if (schemaInfo.getMaximum() != null) {
+            BigDecimal max = new BigDecimal(schemaInfo.getMaximum().toString());
+            boolean inclusive = !Boolean.TRUE.equals(schemaInfo.getExclusiveMaximum());
+            target.addAnnotation(String.format("%s.validation.constraints.DecimalMax", topLevelPackage))
+                .setStringValue(max.toPlainString())
+                .setLiteralValue("inclusive", String.valueOf(inclusive));
+        }
+
+        if (schemaInfo.getMinimum() != null) {
+            BigDecimal min = new BigDecimal(schemaInfo.getMinimum().toString());
+            boolean inclusive = !Boolean.TRUE.equals(schemaInfo.getExclusiveMinimum());
+            target.addAnnotation(String.format("%s.validation.constraints.DecimalMin", topLevelPackage))
+                .setStringValue(min.toPlainString())
+                .setLiteralValue("inclusive", String.valueOf(inclusive));
+        }
+
+        addConstraint(target, schemaInfo.getMaxLength(), sizeConstraint, "max");
+        addConstraint(target, schemaInfo.getMinLength(), sizeConstraint, "min");
+        addConstraint(target, schemaInfo.getPattern(), String.format("%s.Pattern", constraintPkg), "regexp");
+        addConstraint(target, schemaInfo.getDefaultValue(), String.format("%s.ws.rs.DefaultValue", topLevelPackage), "value");
+    }
+
+    <C> void addConstraint(AnnotationTargetSource<?, ?> target, C constraintValue, String annotationName, String annotationProperty) {
+        if (constraintValue instanceof String) {
+            target.addAnnotation(annotationName).setStringValue(annotationProperty, constraintValue.toString());
+        } else if (constraintValue != null) {
+            target.addAnnotation(annotationName).setLiteralValue(annotationProperty, constraintValue.toString());
```

Comment:
I think we should sanitize the `constraintValue` to make sure we are not using reserved keywords or invalid patterns.

Reply:
Are you thinking specifically for `Pattern` constraints or more generally? I think we should be safe from reserved keyword use since the generator is in control of the annotation property names in this part of the code. The danger is present when the `x-codegen-annotations` is used, but ultimately anything invalid (reserved keywords or non-existent annotations/properties) specified by the user would fail to compile.